File: src/data/courses/dsa/two-pointers-sliding-window/sliding-window-fixed-size/subarray-sum-fixed-size-k/code/java/SubarraySumFixedSizeK.java
// ============================================================
// Subarray Sum — Fixed Size K
// ============================================================

// ============================================================
// Solution 1: Brute Force — Recompute Each Window
// Time: O(N×K) | Space: O(1)
// ============================================================
class Solution1 {
    public boolean hasSubarraySum(int[] nums, int k, int target) {
        for (int i = 0; i <= nums.length - k; i++) {
            int sum = 0;
            for (int j = i; j < i + k; j++) sum += nums[j];
            if (sum == target) return true;
        }
        return false;
    }
}

// ============================================================
// Solution 2: Sliding Window — Running Sum (Optimal)
// Time: O(N) | Space: O(1)
// ============================================================
class Solution2 {
    public boolean hasSubarraySum(int[] nums, int k, int target) {
        int sum = 0;
        for (int i = 0; i < k; i++) sum += nums[i];
        if (sum == target) return true;

        for (int i = k; i < nums.length; i++) {
            sum += nums[i] - nums[i - k];
            if (sum == target) return true;
        }

        return false;
    }
}

File: src/data/courses/dsa/two-pointers-sliding-window/sliding-window-fixed-size/subarray-sum-fixed-size-k/code/python/subarray_sum_fixed_size_k.py
# ============================================================
# Subarray Sum — Fixed Size K
# ============================================================

from typing import List

# ============================================================
# Solution 1: Brute Force — Recompute Each Window
# Time: O(N×K) | Space: O(1)
# ============================================================
class Solution1:
    def hasSubarraySum(self, nums: List[int], k: int, target: int) -> bool:
        for i in range(len(nums) - k + 1):
            if sum(nums[i:i+k]) == target:
                return True
        return False

# ============================================================
# Solution 2: Sliding Window — Running Sum (Optimal)
# Time: O(N) | Space: O(1)
# ============================================================
class Solution2:
    def hasSubarraySum(self, nums: List[int], k: int, target: int) -> bool:
        # First window
        window_sum = sum(nums[:k])
        if window_sum == target:
            return True

        # Slide: subtract leaving, add entering
        for i in range(k, len(nums)):
            window_sum += nums[i] - nums[i - k]
            if window_sum == target:
                return True

        return False
